### PR TITLE
Add test for parallel file import on S3

### PR DESCRIPTION
**Pull request by w-gao**
Adds a unittest for #2707.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * Add test for parallel file import on S3

## Reviewer Checklist

 * [x] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [x] Read through the code changes. Make sure that it doesn't have:
    * [x] Addition of trailing whitespace.
    * [x] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [x] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [x] New functions or classes without informative docstrings.
    * [x] Changes to semantics not reflected in the relevant docstrings.
    * [x] New or changed command line options for Toil workflows that are not reflected in `docs/running/cliOptions.rst`
    * [x] New features without tests.
* [x] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [x] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

File: src/toil/test/jobStores/jobStoreTest.py [hunk 1]
```diff
@@ -1339,6 +1339,56 @@ def testOverlargeJob(self):
...
+        """ Tests that importFile is thread-safe."""
+
+        from toil.jobStores.abstractJobStore import AbstractJobStore
+        from concurrent.futures.thread import ThreadPoolExecutor
+        import multiprocessing
+        import timeit
+
+        threads = (
+            2,
+            multiprocessing.cpu_count(),  # maximum
```
Comment (adamnovak) [inline src/toil/test/jobStores/jobStoreTest.py:1352]: We have a Toil-specific CPU counting method (`toil.lib.threading.cpu_count`) that understands Kubernetes pods and won't make as many threads as there are cores on the underlying host.

File: src/toil/test/jobStores/jobStoreTest.py [hunk 2]
```diff
@@ -1339,6 +1339,56 @@ def testOverlargeJob(self):
...
+        testFiles = [other._prepareTestFile(store, size) for _ in range(numOfFiles)]
+
+        def asyncImportFile(jobStore: AbstractJobStore, url: str) -> str:
+            """ Imports the file at the given URL into the jobStore."""
+            logger.info(f'Importing "{url}" into the jobStore...')
+            return jobStore.importFile(url)
+
+        for thread_count in threads:
+            with self.subTest(f'Testing threaded importFile with "{thread_count}" threads.'):
+                # startTime = timeit.default_timer()
```
Comment (adamnovak) [inline src/toil/test/jobStores/jobStoreTest.py:1373]: If you aren't really using timeit you could cut this and the import.

File: src/toil/test/jobStores/jobStoreTest.py [hunk 3]
```diff
@@ -1339,6 +1339,56 @@ def testOverlargeJob(self):
...
+        size = 1 << 16 + 1
+
+        # The string in otherCls() is arbitrary as long as it returns a class that has access
+        # to ._externalStore() and ._prepareTestFile()
+        other = AWSJobStoreTest('testSharedFiles')
+        store = other._externalStore()
+
+        # prepare test files to import
+        logger.info(f'Preparing {numOfFiles} test files for testMultithreadImportFile().')
+        testFiles = [other._prepareTestFile(store, size) for _ in range(numOfFiles)]
```
Comment (adamnovak) [inline src/toil/test/jobStores/jobStoreTest.py:1364]: This should be `snake_case` now.

File: src/toil/test/jobStores/jobStoreTest.py [hunk 4]
```diff
@@ -1339,6 +1339,56 @@ def testOverlargeJob(self):
...
+        from toil.jobStores.abstractJobStore import AbstractJobStore
+        from concurrent.futures.thread import ThreadPoolExecutor
+        import multiprocessing
+        import timeit
+
+        threads = (
+            2,
+            multiprocessing.cpu_count(),  # maximum
+        )
+        numOfFiles = 5
```
Comment (adamnovak) [inline src/toil/test/jobStores/jobStoreTest.py:1354]: This wants to be `snake_case`.

File: src/toil/test/jobStores/jobStoreTest.py [hunk 5]
```diff
@@ -1339,6 +1339,56 @@ def testOverlargeJob(self):
...
+                with ThreadPoolExecutor(max_workers=thread_count) as executor:
+                    for url, expected_md5 in testFiles:
+                        future = executor.submit(asyncImportFile, self.jobstore_initialized, url)
+                        results.append((future, expected_md5))
+
+                # check results
+                self.assertEqual(len(results), numOfFiles)
+
+                for future, expected_md5 in results:
+                    jobStoreFileID = future.result()
```
Comment (adamnovak) [inline src/toil/test/jobStores/jobStoreTest.py:1385]: `snake_case`

File: src/toil/test/jobStores/jobStoreTest.py [hunk 6]
```diff
@@ -1339,6 +1339,56 @@ def testOverlargeJob(self):
...
+                    self.assertIsInstance(jobStoreFileID, FileID)
+
+                    with self.jobstore_initialized.readFileStream(jobStoreFileID) as f:
+                        fileMD5 = hashlib.md5(f.read()).hexdigest()
```
Comment (adamnovak) [inline src/toil/test/jobStores/jobStoreTest.py:1389]: `snake_case`

**Review by adamnovak — CHANGES_REQUESTED** [review]
This looks like it should work, but it does seem to need some cleanup.

**Comment by adamnovak** [comment]
I think the Gitlab tests might have only failed for Kubernetes being full and unable to run stuff for us promptly.

File: src/toil/test/jobStores/jobStoreTest.py [hunk 7]
```diff
@@ -1339,18 +1340,61 @@ def testOverlargeJob(self):
...
+                    with self.jobstore_initialized.readFileStream(file_id) as f:
+                        file_md5 = hashlib.md5(f.read()).hexdigest()
+                    self.assertEqual(file_md5, expected_md5)
+
     def _prepareTestFile(self, bucket, size=None):
         from toil.jobStores.aws.utils import retry_s3
 
-        fileName = 'testfile_%s' % uuid.uuid4()
-        url = 's3://%s/%s' % (bucket.name, fileName)
+        file_name = 'testfile_%s' % uuid.uuid4()
```
Comment (DailyDreaming) [inline src/toil/test/jobStores/jobStoreTest.py:1389]: I appreciate the value-added snake case.

**Review by DailyDreaming — COMMENTED** [review]
Overall, looks good to me.  See minor comments.

**Comment by adamnovak** [comment]
@w-gao Is this PR ready to not be a draft anymore?

**Comment by w-gao** [comment]
@adamnovak I think so, though I still have to update the type hints so diff_mypy can pass.

**Review by DailyDreaming — APPROVED** [review]
@w-gao Typing is pretty important for working code, but I'm of the opinion that typing in pure test files is nice to have, but not critical.  I'm in favor of merging this if tests pass.

**Review by adamnovak — APPROVED** [review]

**Comment by DailyDreaming** [comment]
Tests have passed, I'm thinking that this is good to merge.  Thank you @w-gao .

**Pull request merged**